### ISSUE TITLE
Repeating parts without PDF code replication

### ISSUE THREAD

**Issue by MichaelFraiman** [typst/typst#3712]
### Description

This issue addresses the request for a functionality similar to the one provided by the `xsavebox` package for LaTeX, see https://ctan.org/pkg/xsavebox.
Succinctly, the idea is that a certain "constant" content is created and then placed into the document multiple times without actually replicating the PDF code. This allows the document to contain repeating parts without increasing in size.

Syntax might be something like this

```
#create_box(name: "mybox")[contents]
for i in range(100) {
#pagebreak()
#use_box("mybox")
}
```

It also might be useful to allow for reference points to make insertion of some other objects more simple (see use cases).

### Use Case

The following functionality can be used to produce documents with a lot of repeating content, such as repeating headers or footers, repeating background. Also it can be used to create personalised handouts, which contain only a few features that differ one from the other.
For example, some form intended for printing, where the name of the person is printed:
```
Name: ______
*large picture*
*some text* *personalised QR-code*
*another picture*
```

**Comment by laurmaedje** [comment]
This sounds like something Typst should transparently to optimize file size, without any user intervention. 

Also of note: If the objects are on the same page, then the compression algorithm would just take care of removing the duplication. With compressed object streams, we could extend this to the whole file. (Possibly at the cost of rendering performance?)

**Comment by reknih** [comment]
I don't think that this is something the user should be exposed to. Instead, if at all, Typst should automatically detect duplicate content streams and reference them as XObjects.

A good alternative could be the use of _Object Streams_ (PDF 1.7 spec, 7.5.7) which can contain arbitrary PDF and then be compressed with deflate. This would identify and compress identical graphics streams and other objects in a size-optimized PDF writing profile.

**Comment by MichaelFraiman** [comment]
I would say it should still be possible to do manually

**Comment by jvoisin** [comment]
> I would say it should still be possible to do it manually.

Why?
